Added option to delete or preserve k8s pods after completion

nuwang opened the pull request:
This PR adds an option to delete job pods after running to completion. It follows the conventions of the `cleanup_job` option in the galaxy config, and provides options between `always`, `onsuccess` and `never`.

nuwang committed:
Added option to delete or preserve k8s pods after completion

nuwang committed:
Send DeleteOptions when deleting k8s jobs

mvdbeek committed:
Fix watch_checked_item exception when deleting job externally

Avoids:
```
galaxy.jobs.runners.kubernetes ERROR 2019-06-16 19:06:43,961 No Jobs are available under expected selector app=galaxy-gx-short-id-5
galaxy.jobs.runners ERROR 2019-06-16 19:06:43,962 Unhandled exception checking active jobs
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/runners/__init__.py", line 696, in monitor
    self.check_watched_items()
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/runners/__init__.py", line 723, in check_watched_items
    new_async_job_state = self.check_watched_item(async_job_state)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/runners/kubernetes.py", line 447, in check_watched_item
    with open(job_state.error_file, 'w') as error_file:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/mvandenb/temp/tmp1nx5r6_s/000/5/galaxy_5.e'
```
Can be tested with the `test_kill_process` testcase.

mvdbeek committed:
Use set literal

mvdbeek committed:
Adjust testcase for new delete job behavior

mvdbeek commented:
Thanks a lot @nuwang , I've added some minor things in https://github.com/nuwang/galaxy/pull/3, with those in place I'm +1 on the PR!

nuwang commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -465,9 +474,25 @@ def _handle_job_failure(self, job, job_state, reason=None):
 
+    def __cleanup_k8s_job(self, job):
+        k8s_cleanup_job = self.runner_params['k8s_cleanup_job']
+        job_failed = (job.obj['status']['failed'] > 0
```
@mvdbeek The `job_failed` test on this line only checks whether the k8s job pod has failed, not whether the galaxy job has failed. I suppose we should also leave the k8s job behind if the Galaxy job fails? Also, another, related, issue is that the k8s pods do not have any stdout or stderr, as the output is being redirected to files. Therefore, pods that are left behind offer little in terms of debugging potential. It also kind of goes against the grain of k8s, where pod logs could be coupled with a logstash/kibana combo for tracking job failures. Of course, I suppose the job logs could be obtained from Galaxy and sent to logstash/kibana too. What are your thoughts on this?

mvdbeek replied:
> Also, another, related, issue is that the k8s pods do not have any stdout or stderr, as the output is being redirected to files.

Are you sure about that ? I had modified the runner specifically for that. If they're being swallowed now we can surely recover them.

nuwang replied:
So far I've not seen any output at all on any of the pods, regardless of success or failure. I assumed that was because the galaxy job script was redirecting output to files, and I didn't spot a `tee` or anything that redirected output to both. Is there some option I need to set?

mvdbeek replied:
It should happen here: https://github.com/galaxyproject/galaxy/blob/bf31260192eede738e5e9e41b1cd1b490449333b/lib/galaxy/jobs/command_factory.py#L257 but we may need to adapt that to the new stdout/stderr collection strategy. I can add a test for this.

nuwang replied:
That's great, I thought I saw somewhere that was supposed to happen, so that clarifies things. Should I also add an explicit check to avoid cleanup if galaxy determines the job as failed? I kind of recall that some tools incorrectly write errors to stdout which Galaxy explicitly detects?

mvdbeek replied:
I'd say no. If the job hasn't been killed by the runner and the error is instead inferred by some other condition like presence of stderr I don't think there'd be any value in keeping the kubernetes job around.

nuwang commented:
@martenson @mvdbeek I just realised that this PR was never merged into 19.09 although it was reviewed and approved because of some conflicts. I've rebased and fixed the conflicts, can we still merge this into 19.09?

martenson commented:
I'll let Marius re-evaluate but see no problem in getting this in. Changed milestone to 19.09.

mvdbeek commented:
Nice, thanks @nuwang!
